src/connectivity.jl

```diff
@@ -217,102 +217,159 @@ julia> strongly_connected_components(g)
  [8, 9]
  [5, 6, 7]
  [1, 2, 3, 4]
- [10, 11]
+ [10, 11]    
+
+
+This currently uses a modern variation on Tarjan's algorithm, largely derived from algorithm 3 in David J. Pearce's 
+preprint: https://homepages.ecs.vuw.ac.nz/~djp/files/IPL15-preprint.pdf   , with some changes & tradeoffs when unrolling it to an
+imperative algorithm.
 ```
 """
 function strongly_connected_components end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
+
 @traitfn function strongly_connected_components(
     g::AG::IsDirected
 ) where {T<:Integer,AG<:AbstractGraph{T}}
-    zero_t = zero(T)
-    one_t = one(T)
-    nvg = nv(g)
-    count = one_t
+    if iszero(nv(g))
+        return Vector{Vector{T}}()
+    end
+    return _strongly_connected_components_tarjan(g, infer_nb_iterstate_type(g))
+end
+
+#  In recursive form, Tarjans algorithm has a recursive call inside a for loop.
+#  To save the loop state of each recursive step in a stack efficiently, 
+#  we need to infer the type of its state (which should almost always be an int).
+destructure_type(x) = Any
+destructure_type(::Type{Union{Nothing,Tuple{<:Any,B}}}) where {B} = B
```

Comment:
Aqua complains because this is ill-defined: `Union{A,B}` is the same as `Union{B,A}`

Reply:
This seems needlessly convoluted, is it really needed?

Reply:
The thing that is actually needed is the infer_nb_iterstate_type below. You could just make that return Any for the non-AbstractSimpleGraph case, and overload the is_large_vertex function to a much higher threshold to avoid regressions for user-defined graph types.

Reply:
The thing that the function does is fairly simple though, it just unpacks the type of the objects returned by an iterator. Referring to that type would be straightforward to do in any statically typed language where iterators have a typed interface instead of being a duck typed protocol, but is slightly tricky in Julia

Reply:
`outneighbors` should return vertices of the graph, which should be of type `eltype(g)`, why do you need to use `infer_nb_iterstate_type` method, where you can just return `T`?

Reply:
eltype is the first tuple element, I need the second (usually an Int, but could be a linked list node for example), so that when outneighbours is suspended I can restart from the same location instead of restarting from the beginning and making the algorithm quadratic in the number of edges.